system/jlib/jqueue.tpp

```diff
@@ -547,4 +547,98 @@ public:
                                     while (x--) 
 
 
+template <class BASE>
+class RingBufferOf
+{
+    typedef RingBufferOf<BASE> SELF;
+    BASE **ptrs;
+    unsigned headp;
+    unsigned tailp;
+    unsigned max;
+    unsigned num;
+
+    virtual void releaseBuffer(BASE *e){};
+    virtual void releaseBufferAll()
+    {
+        while (num)
+        {
+            releaseBuffer(ptrs[headp]);
+            headp++;
+            num--;
+            if (headp==max) headp=0;
+        }
+    };
+
+public:
+    inline RingBufferOf() : ptrs(NULL), headp(0), tailp(0), max(0), num(0){};
+
+    virtual ~RingBufferOf() { clear(); }
+
+    inline void clear()
+    {
+        releaseBufferAll();
+        free(ptrs);
+        ptrs=NULL;
+        headp=0;
+        tailp=0;
+        max=0;
+        num=0;
+    }
+    inline void ensure(unsigned bufSize)
+    {
+        clear();
+        max = bufSize;
+        ptrs = new BASE* [max];
+        assertex(!max || ptrs);
+    }
+    inline void enqueue(BASE *e)
+    {
+        if (e) {
+            ++tailp;
+            ++num;
+            if (tailp==max)
+                tailp = 0;
+            if (headp==tailp)
+            {
+                releaseBuffer(ptrs[headp]);
+                headp++;
+                if (headp==max)
+                    headp=0;
+                num--;
+            }
+            ptrs[tailp] = e;
+        }
+    }
+    inline BASE *dequeue()
+    {
+        if (!num)
+            return NULL;
+        BASE *ret = ptrs[headp];
+        headp++;
+        if (headp==max)
+            headp=0;
+        num--;
+        return ret;
+    }
+    inline unsigned ordinality() const { return num; }
+};
+
+template <class BASE>
+class SafeRingBufferOf : private RingBufferOf<BASE>
+{
+   typedef SafeRingBufferOf<BASE> SELF;
+protected:
+    mutable CriticalSection crit;
+    inline void unsafeenqueue(BASE *e) { RingBufferOf<BASE>::enqueue(e); }
+    inline BASE *unsafedequeue() { return RingBufferOf<BASE>::dequeue(); }
+    inline unsigned unsafeordinality() { return RingBufferOf<BASE>::ordinality(); }
+
+public:
+    inline void ensure(unsigned bufSize) { RingBufferOf<BASE>::ensure(bufSize);}
+    void clear() { CriticalBlock b(crit); RingBufferOf<BASE>::clear(); }
+    void enqueue(BASE *e) { CriticalBlock b(crit); RingBufferOf<BASE>::enqueue(e); }
+    BASE *dequeue() { CriticalBlock b(crit); return RingBufferOf<BASE>::dequeue(); }
+    inline unsigned ordinality() const { CriticalBlock b(crit); return RingBufferOf<BASE>::ordinality(); }
```

Comment:
This is a good first approximation for a ring buffer implementation, but I am a bit concerned about the efficiency.  In particular grabbing a critical section each time a row is added is potentially painful.

As part of phase two we should consider a lock free implementation.  It may be tricky, (I have a couple of ideas), but would potentially save a lot of time.